src/Lut.ts

```diff
@@ -0,0 +1,500 @@
+ * Need to be converted to LUT for rendering.
+ * @property {number} x The X Coordinate
+ * @property {number} opacity The Opacity, from 0 to 1
+ * @property {Array.<number>} color The Color, 3 numbers from 0-255 for r,g,b
+ */
+export type ControlPoint = {
+  x: number;
+  opacity: number;
+  color: [number, number, number];
+};
+
+function controlPointToRGBA(controlPoint) {
+  return [controlPoint.color[0], controlPoint.color[1], controlPoint.color[2], Math.floor(controlPoint.opacity * 255)];
+}
+
+/**
+ * @typedef {Object} Lut Used for rendering.
+ * @property {Array.<number>} lut LUT_ARRAY_LENGTH element lookup table as array
+ * (maps scalar intensity to a rgb color plus alpha, with each value from 0-255)
+ * @property {Array.<ControlPoint>} controlPoints
+ */
+export class Lut {
+  public lut: Uint8Array;
+  public controlPoints: ControlPoint[];
+
+  constructor() {
+    this.lut = new Uint8Array(LUT_ARRAY_LENGTH);
+    this.controlPoints = [];
+    this.createFullRange();
+  }
+
+  /**
+   * Generate a piecewise linear lookup table that ramps up from 0 to 1 over the b to e domain.
+   * If e === b, then we use a step function with f(b) = 0 and f(b + 1) = 1
+   *  |
+   * 1|               +---------+-----
+   *  |              /
+   *  |             /
+   *  |            /
+   *  |           /
+   *  |          /
+   * 0+=========+---------------+-----
+   *  0         b    e         255
+   * @return {Lut}
+   * @param {number} b
+   * @param {number} e
+   */
+  createFromMinMax(b: number, e: number): Lut {
+    if (e < b) {
+      // swap
+      const tmp = e;
+      e = b;
+      b = tmp;
+    }
+    const lut = new Uint8Array(LUT_ARRAY_LENGTH);
+    for (let x = 0; x < lut.length / 4; ++x) {
+      lut[x * 4 + 0] = 255;
+      lut[x * 4 + 1] = 255;
+      lut[x * 4 + 2] = 255;
+      if (x > e) {
+        lut[x * 4 + 3] = 255;
+      } else if (x <= b) {
+        lut[x * 4 + 3] = 0;
+      } else {
+        if (e === b) {
+          lut[x * 4 + 3] = 255;
+        } else {
+          const a = (x - b) / (e - b);
+          lut[x * 4 + 3] = lerp(0, 255, a);
+        }
+      }
+    }
+
+    // Edge case: b and e are both out of bounds
+    if (b < 0 && e < 0) {
+      this.lut = lut;
+      this.controlPoints = [
+        { x: 0, opacity: 1, color: [255, 255, 255] },
+        { x: 255, opacity: 1, color: [255, 255, 255] },
+      ];
+      return this;
+    }
+    if (b >= 255 && e >= 255) {
+      this.lut = lut;
+      this.controlPoints = [
+        { x: 0, opacity: 0, color: [255, 255, 255] },
+        { x: 255, opacity: 0, color: [255, 255, 255] },
+      ];
+      return this;
+    }
+
+    // Generate 2 to 4 control points for a minMax LUT, from left to right
+    const controlPoints: ControlPoint[] = [];
+
+    // Add starting point at x = 0
+    let startVal = 0;
+    if (b < 0) {
+      startVal = -b / (e - b);
+    }
+    controlPoints.push({ x: 0, opacity: startVal, color: [255, 255, 255] });
```

Comment:
For my understanding: if `b` is negative -10 and e is 20, the starting value is the ratio 10/30, so the opacity of the first control point is 0.33. So assuming that `e` is inside the range normalized to 255, this lut will range from 30-100% opacity?

If an input value is less than -10 will it get 0.33 for opacity? Or will it get 0?

Reply:
You have it right.  If b is `-10` and e is `20`, then the lookup table will start at 0.33 at x=0, rise up linearly to 1.0 at x=20, and then stay flat at 1.0 all the way to x=255.

Input values to the lookup table are clamped to 0..255.  So it will never sample outside of those bounds, so you'll always get numbers from 0.33 up to 1.0.